### PR TITLE
Feature/user permissions

### PR DESCRIPTION
**Pull request by mateuszfelczak**
This feature adds a method has_permission to the User class. Permission type is checked at first (there's only 1 permission type now but it can be extended). If the method is used on the post author, community moderator or a community admin, and the permission type is "edit", then it returns True.

## Summary by Sourcery

Introduce a new has_permission method in the User class to check user permissions for editing posts. Refactor UserManager methods for better readability. Add tests to ensure the correct functionality of the new permission checks.

New Features:
- Add has_permission method to the User class to check if a user has the 'edit' permission for a post.

Enhancements:
- Refactor UserManager methods to improve readability by reformatting the method signatures.

Tests:
- Introduce test_permissions.py to validate the has_permission method with various user roles and scenarios.

### CODE REVIEW

**Comment by jacoor** [comment]
@mateuszfelczak fix the tests, update the branch. Only then in can be checked.

**Comment by jacoor** [comment]
@mateuszfelczak fix the tests, update the branch. Only then in can be checked.

**Comment by jacoor** [comment]
@mateuszfelczak how is it going with fixes?

**Comment by jacoor** [comment]
@mateuszfelczak please fix linting. If you need help, ping me on Discord. 
As you see here, https://github.com/pymasterspl/reddit/actions/runs/10274408032/job/28431057540?pr=73 there are formatting issues. 
Instructions on how to format automatically are in readme.

**Comment by jacoor** [comment]
Also, @mateuszfelczak first line of the PR description should contain link to clickup task.

File: core/models.py [hunk 1]
```diff
@@ -37,8 +36,10 @@ class Meta:
 class Community(GenericModel):
     name = models.CharField(max_length=255)
     slug = models.SlugField(max_length=255, unique=True, blank=True)
-    author = models.ForeignKey(User, on_delete=models.SET_NULL, null=True, related_name="authored_communities")
-    members = models.ManyToManyField(User, through="CommunityMember", related_name="communities")
+    author = models.ForeignKey(
```
Comment (jacoor) [inline core/models.py]: Revert this changes please. Seems your editor is set to shorter lines than project settings, you need to adjust. 
Our line lenght is 120 chars: https://github.com/pymasterspl/reddit/blob/6810df3c26114963e2be001cad26822025886179/pyproject.toml#L33

The problem: this change breakes git history and makes debugging harder also CR longer.

File: core/models.py [hunk 2]
```diff
@@ -61,7 +62,8 @@ def generate_unique_slug(self: "Community") -> str:
         return unique_slug
 
     def count_online_users(self: "Community") -> int:
-        online_limit = timezone.now() - timedelta(minutes=settings.LAST_ACTIVITY_ONLINE_LIMIT_MINUTES)
```
Comment (jacoor) [inline core/models.py]: Revert this changes please. Seems your editor is set to shorter lines than project settings, you need to adjust. 
Our line lenght is 120 chars: https://github.com/pymasterspl/reddit/blob/6810df3c26114963e2be001cad26822025886179/pyproject.toml#L33

The problem: this change breaks git history and makes debugging harder also CR longer. 

Please revert all those line length changes and re request review.

**Review by jacoor — CHANGES_REQUESTED** [review]

**Comment by jacoor** [comment]
@sourcery-ai review

File: users/models.py [hunk 3]
```diff
@@ -16,7 +17,13 @@
 class UserManager(BaseUserManager):
     use_in_migrations: bool = True
 
-    def _create_user(self: "UserManager", email: str, nickname: str, password: str, **extra_fields: dict) -> "User":
+    def _create_user(
```
Comment (jacoor) [inline users/models.py]: Please revert - formatting.

File: users/models.py [hunk 4]
```diff
@@ -29,13 +36,25 @@ def _create_user(self: "UserManager", email: str, nickname: str, password: str,
         user.save(using=self._db)
         return user
 
-    def create_user(self: "UserManager", email: str, nickname: str, password: str, **extra_fields: dict) -> "User":
+    def create_user(
```
Comment (jacoor) [inline users/models.py]: Please revert - formatting.

File: users/models.py [hunk 5]
```diff
@@ -29,13 +36,25 @@ def _create_user(self: "UserManager", email: str, nickname: str, password: str,
...
+        nickname: str,
+        password: str,
+        **extra_fields: dict,
+    ) -> "User":
         extra_fields.setdefault("is_staff", False)
         extra_fields.setdefault("is_superuser", False)
 
         return self._create_user(email, nickname, password, **extra_fields)
 
-    def create_superuser(self: "UserManager", email: str, nickname: str, password: str, **extra_fields: dict) -> "User":
```
Comment (jacoor) [inline users/models.py]: Please revert - formatting.

File: users/models.py [hunk 6]
```diff
@@ -78,6 +99,37 @@
...
+        Post = apps.get_model("core", "Post")
+        CommunityMember = apps.get_model("core", "CommunityMember")
+
+        try:
+            post = Post.objects.get(id=post_id)
+        except Post.DoesNotExist:
+            print(f"Post with id {post_id} does not exist.")
+            return False
+
+        match permission_name:
```
[the remark this answers is not in the document]
Comment (jacoor) [reply users/models.py]: @mateuszfelczak AI suggestion here is pretty good but I would do it bit different way.
Instead of using lambda, move the code from line 111 to 120 into a function, called, eg.
__check_permission_post_edit (two underscores for private method)
Then do  (pseudocode)
```
        permission_checkers = {
            "edit": __check_permission_post_edit(PARAMETERS)
            # Add more permissions here as needed
        }
```
this way we have really clean has_permission method in models and __check_permission_post_edit is separate function. This again, gives us the option to write tests only for __check_permission_post_edit, not for the whole has_permssion thing. So please also rewrite your already good test into much better ones
- all the tests you have for has_permission can be done on __check_permission_post_edit
- please add two tests (positive and negative) for user.has_has_permission with the edit permission. This willl be integration tests showing that __check_permission_post_edit is actually used. 
-- Or you can get really fancy and try to use spy while calling has_permission to show that __check_permission_post_edit was called. Code example below.


==== Using mock (spy) to show function was called ===

```python
from unittest.mock import MagicMock

def some_function():
    pass

def function_to_test():
    some_function()

def test_function_to_test():
    mock = MagicMock()
    globals()['some_function'] = mock
    function_to_test()
    mock.assert_called_once()

if __name__ == "__main__":
    import pytest
    pytest.main()
```

In this example:

1. Create a `MagicMock` for `some_function`.
2. Replace the original `some_function` with the mock.
3. Call `function_to_test`.
4. Verify that `some_function` was called exactly once using `assert_called_once`.

This is a simple and effective way to verify function calls in your tests. Happy testing!

File: users/tests/test_permissions.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
+import pytest
+from django.contrib.auth.hashers import make_password
+
+from core.models import Community, CommunityMember, Post
+from users.models import User
+
+
+@pytest.fixture()
+def user() -> User:
```
Comment (jacoor) [inline users/tests/test_permissions.py]: not needed:
https://github.com/pymasterspl/reddit/blob/6810df3c26114963e2be001cad26822025886179/conftest.py#L30

File: users/tests/test_permissions.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    return User.objects.create_user(email="user@example.com", nickname="user", password=make_password("password"))
+
+
+@pytest.fixture()
+def user_without_post() -> User:
```
Comment (jacoor) [inline users/tests/test_permissions.py]: it's the same as above no? So please remove and use only one from conftests.py

File: users/tests/test_permissions.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    )
+
+
+@pytest.fixture()
+def admin() -> User:
+    return User.objects.create_user(email="admin@example.com", nickname="admin", password=make_password("password"))
+
+
+@pytest.fixture()
+def community(user: User) -> Community:
```
Comment (jacoor) [inline users/tests/test_permissions.py]: not needed, please use 
https://github.com/pymasterspl/reddit/blob/6810df3c26114963e2be001cad26822025886179/conftest.py#L55-L56

File: users/tests/test_permissions.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    return Community.objects.create(name="Test Community", author=user)
+
+
+@pytest.fixture()
+def post(user: User, community: Community) -> Post:
```
Comment (jacoor) [inline users/tests/test_permissions.py]: Not needed, please use 
https://github.com/pymasterspl/reddit/blob/6810df3c26114963e2be001cad26822025886179/conftest.py#L60-L61

File: users/tests/test_permissions.py [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    assert user.has_permission(post_id=post.id, permission_name="edit") is True
+
+
+@pytest.mark.django_db()
+def test_user_cannot_edit_others_post(post: Post, user_without_post: User) -> None:
+    assert user_without_post.has_permission(post_id=post.id, permission_name="edit") is False
+
+
+@pytest.mark.django_db()
+def test_moderator_can_edit_post(moderator: User, post: Post, community_member_moderator: CommunityMember) -> None:
```
Comment (jacoor) [inline users/tests/test_permissions.py]: Would be good to see if moderator can not edit posts on other communities. Admin probably too.

**Review by jacoor — CHANGES_REQUESTED** [review]

File: users/tests/test_permissions.py [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+def community_member_moderator(moderator: User, community: Community) -> CommunityMember:
+    return CommunityMember.objects.create(user=moderator, community=community, role=CommunityMember.MODERATOR)
+
+
+@pytest.fixture()
+def community_member_admin(admin: User, community: Community) -> CommunityMember:
+    return CommunityMember.objects.create(user=admin, community=community, role=CommunityMember.ADMIN)
+
+
+@pytest.mark.django_db()
```
[the remark this answers is not in the document]
Comment (jacoor) [reply users/tests/test_permissions.py:46]: Looks like free tests. Please implement (commit suggestion  button works nicely)

File: users/models.py [hunk 13]
```diff
@@ -16,7 +17,13 @@
...
     use_in_migrations: bool = True
 
-    def _create_user(self: "UserManager", email: str, nickname: str, password: str, **extra_fields: dict) -> "User":
+    def _create_user(
+        self: "UserManager",
+        email: str,
+        nickname: str,
+        password: str,
+        **extra_fields: dict,
+    ) -> "User":
```
Comment (jacoor) [inline users/models.py:26]:
```suggestion
        self: "UserManager",
        email: str,
```

File: conftest.py [hunk 14]
```diff
@@ -46,7 +46,11 @@ def another_user(generated_password: str) -> User:
 
 @pytest.fixture()
 def admin_user(generated_password: str) -> User:
-    admin_user = User.objects.create_superuser(email="admin@example.com", nickname="admin", password=generated_password)
+    admin_user = User.objects.create_superuser(
```
Comment (jacoor) [inline conftest.py:56]: Revert this formatting change, please.

File: users/models.py (hunk 3 above)
Comment (jacoor) [reply users/models.py]: Still, needs to be reverted.

File: users/models.py (hunk 4 above)
Comment (jacoor) [reply users/models.py]: Again, revert.

File: users/models.py (hunk 5 above)
Comment (jacoor) [reply users/models.py]: revert.

File: users/tests/test_permissions.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+def test_check_permission_post_edit_admin(
+    admin_user: User, post: Post, community_member_admin: CommunityMember
+) -> None:
+    assert admin_user._User__check_permission_post_edit(post.id) is True  # noqa: SLF001
+
+
+@pytest.mark.django_db()
+def test_check_permission_post_edit_not_an_admin(admin_user: User, post2: Post) -> None:
+    assert (
```
Comment (jacoor) [inline users/tests/test_permissions.py]: This does not need brackets, this line is shorter than 120 chars. I did the count. Please fix your editor.

File: users/tests/test_permissions.py [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+        admin_user._User__check_permission_post_edit(post2.id) is False  # noqa: SLF001
+    )
+
+
+@pytest.mark.django_db()
+def test_check_permission_post_edit_user_not_post_owner(another_user: User, post: Post) -> None:
+    assert (
```
Comment (jacoor) [inline users/tests/test_permissions.py]: This does not need brackets, this line is shorter than 120 chars. I did the count. Pleae fix your editor.

File: users/tests/test_permissions.py [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+
+@pytest.mark.django_db()
+def test_check_permission_post_edit_non_existent_post(user: User) -> None:
+    assert user._User__check_permission_post_edit(99999) is False  # noqa: SLF001
+
+
+@pytest.mark.django_db()
+def test_has_permission_calls_check_permission_post_edit(user: User, post: Post) -> None:
+    mock = MagicMock()
+    user._User__check_permission_post_edit = mock  # noqa: SLF001
```
Comment (jacoor) [inline users/tests/test_permissions.py:71]: This is nice!

**Review by jacoor — CHANGES_REQUESTED** [review]
You still need to revert not needed changes.

**Comment by jacoor** [comment]
@mateuszfelczak how is it going? Need help?

**Comment by jacoor** [comment]
@mateuszfelczak how is it going with this? Do you need help?

**Review by jacoor — APPROVED** [review]
